src/digits/ff31.rs

```diff
@@ -0,0 +1,1373 @@
+            use $crate::digits::constant_bool::*;
+            use $crate::digits::constant_time_primitives::*;
+            use $crate::digits::ff31::convert_bytes_to_limbs_mut;
+
+            pub const LIMBSIZEBITS: usize = 31;
+            pub const BITSPERBYTE: usize = 8;
+            pub const PRIME: [u32; NUMLIMBS] = $prime;
+            pub const PRIMEBITS: usize = $bits;
+            pub const PRIMEBYTES: usize = PRIMEBITS / BITSPERBYTE;
+            pub const NUMLIMBS: usize = $limbs;
+            pub const NUMDOUBLELIMBS: usize = $limbs * 2;
+            pub const BARRETTMU: [u32; NUMLIMBS + 1] = $barrettmu;
+            pub const MONTRINV: [u32; NUMLIMBS] = $montgomery_r_inv;
+            pub const MONTRSQUARED: [u32; NUMLIMBS] = $montgomery_r_squared;
+            pub const MONTM0INV: u32 = $montgomery_m0_inv;
+
+            #[derive(PartialEq, Eq, Ord, Clone, Copy)]
+            pub struct $classname {
+                pub(crate) limbs: [u32; NUMLIMBS],
+            }
+
+            #[derive(PartialEq, Eq, Ord, Clone, Copy)]
+            pub struct Monty {
+                pub(crate) limbs: [u32; NUMLIMBS],
+            }
+
+            pub struct FpBitIter<'a, $classname: 'a> {
+                p: *const $classname,
+                index: usize,
+                endindex: usize,
+                _marker: marker::PhantomData<&'a $classname>,
+            }
+
+            impl<'a> Iterator for FpBitIter<'a, $classname> {
+                type Item = ConstantBool<u32>;
+                #[inline]
+                fn next(&mut self) -> Option<Self::Item> {
+                    self.index += 1;
+                    let limbs = unsafe { (*self.p).limbs };
+                    if self.index <= self.endindex {
+                        Some($classname::test_bit(&limbs, self.index - 1))
+                    } else {
+                        None
+                    }
+                }
+            }
+
+            impl<'a> DoubleEndedIterator for FpBitIter<'a, $classname> {
+                #[inline]
+                fn next_back(&mut self) -> Option<ConstantBool<u32>> {
+                    self.endindex -= 1;
+                    let limbs = unsafe { (*self.p).limbs };
+                    if self.index <= self.endindex {
+                        Some($classname::test_bit(&limbs, self.endindex))
+                    } else {
+                        None
+                    }
+                }
+            }
+
+            impl fmt::Debug for $classname {
+                fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+                    write!(f, "{}(", stringify!($classname))?;
+                    let x: Vec<String> = self.limbs.iter().map(|x| format!("{:#x}", x)).collect();
+                    write!(f, "{}", x.join(", "))?;
+                    write!(f, ")")?;
+                    Ok(())
+                }
+            }
+
+            // impl fmt::Display for $classname {
```

Comment:
Not sure if this code works or not. If it does, is there any downside to uncommenting it?